sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosPermissionProperties.java

```diff
@@ -45,25 +57,102 @@ public CosmosPermissionProperties setId(String id) {
      */
     CosmosPermissionProperties(String jsonString) {
         this.permission = new Permission(jsonString);
+        this.permissionName = permission.getId();
+        this.permissionMode = permission.getPermissionMode();
+        this.resourcePartitionKey = permission.getResourcePartitionKey();
+        this.resourceToken = permission.getToken();
+
+        String[] parts = Utils.trimBeginningAndEndingSlashes(permission.getResourceLink()).split("/");
+
+        if (parts.length < 4) {
+            throw new IllegalArgumentException("jsonString");
+        }
+
+        this.databaseName = parts[1];
+        this.containerName = parts[3];
+
+        if (parts.length >= 6) {
+            this.resourceName = parts[5];
+
+            if (Paths.DOCUMENTS_PATH_SEGMENT.equalsIgnoreCase(parts[4])) {
+                this.resourceKind = CosmosContainerChildResourceKind.ITEM;
+            } else if (Paths.STORED_PROCEDURES_PATH_SEGMENT.equalsIgnoreCase(parts[4])) {
+                this.resourceKind = CosmosContainerChildResourceKind.STORED_PROCEDURE;
+            } else if (Paths.USER_DEFINED_FUNCTIONS_PATH_SEGMENT.equalsIgnoreCase(parts[4])) {
+                this.resourceKind = CosmosContainerChildResourceKind.USER_DEFINED_FUNCTION;
+            } else if (Paths.TRIGGERS_PATH_SEGMENT.equalsIgnoreCase(parts[4])) {
+                this.resourceKind = CosmosContainerChildResourceKind.TRIGGER;
+            } else {
+                throw new IllegalArgumentException("jsonString");
+            }
+        }
     }
 
     /**
-     * Gets the self-link of resource to which the permission applies.
+     * Sets the name of the Cosmos container as the parent resource which is associated with this permission object.
      *
-     * @return the resource link.
+     * @param containerName the name of the Cosmos container representing the parent resource.
+     * @return the current {@link CosmosPermissionProperties} object.
      */
-    public String getResourceLink() {
-        return this.permission.getResourceLink();
+    public CosmosPermissionProperties setContainerName(String containerName) {
+        this.containerName = containerName;
+
+        // Following is required by permission replace (PUT) scenario.
+        if (this.permission != null) {
+            this.permission.setResourceLink(this.databaseName);
+            this.resourceToken = null;
+        }
+
+        return this;
     }
 
     /**
-     * Sets the self-link of resource to which the permission applies.
+     * Gets the name of the Cosmos container as the parent resource which is associated with this permission object.
      *
-     * @param resourceLink the resource link.
-     * @return the current {@link CosmosPermissionProperties} object
+     * @return the name of the Cosmos container representing the parent resource.
+     */
+    public String getContainerName() {
```

Comment:
this is called containerName but in the test `.setContainerName("AQAAAJ0fgTc=")` uses portion of the selflink.
So is this the container name or just a portion of the selflink?